File: Examples/StereoKitCTest/demo_postfx.cpp
#include "demo_postfx.h"

#include <stereokit.h>
#include <stereokit_ui.h>

#include "skt_postfx_vignette.hlsl.h"
#include "skt_postfx_invert.hlsl.h"
#include "skt_postfx_depth_fog.hlsl.h"

using namespace sk;

///////////////////////////////////////////

static shader_t   postfx_vignette_shader = {};
static shader_t   postfx_invert_shader   = {};
static shader_t   postfx_fog_shader      = {};
static material_t postfx_vignette        = {};
static material_t postfx_invert          = {};
static material_t postfx_fog             = {};
static model_t    postfx_model           = {};
static bool32_t   vignette_on            = false;
static bool32_t   invert_on              = false;
static bool32_t   fog_on                 = false;
static float      vignette_strength      = 0.4f;
static float      fog_density            = 0.5f;

///////////////////////////////////////////

void demo_postfx_init() {
	postfx_vignette_shader = shader_create_mem((void*)sks_skt_postfx_vignette_hlsl,  sizeof(sks_skt_postfx_vignette_hlsl));
	postfx_invert_shader   = shader_create_mem((void*)sks_skt_postfx_invert_hlsl,    sizeof(sks_skt_postfx_invert_hlsl));
	postfx_fog_shader      = shader_create_mem((void*)sks_skt_postfx_depth_fog_hlsl, sizeof(sks_skt_postfx_depth_fog_hlsl));
	postfx_vignette        = material_create(postfx_vignette_shader);
	postfx_invert          = material_create(postfx_invert_shader);
	postfx_fog             = material_create(postfx_fog_shader);

	postfx_model = model_create_file("DamagedHelmet.gltf");
}

///////////////////////////////////////////

// Chain order is the array order - fog reads scene depth first, then the
// invert flips colors, and the vignette darkens the corners last.
static void postfx_apply_chain() {
	material_t chain[3];
	int32_t    count = 0;
	if (fog_on     ) chain[count++] = postfx_fog;
	if (invert_on  ) chain[count++] = postfx_invert;
	if (vignette_on) chain[count++] = postfx_vignette;
	render_set_post_process(chain, count);
}

///////////////////////////////////////////

void demo_postfx_update() {
	// Helmet on the left, settings on the right, centered together in front.
	model_draw(postfx_model, matrix_trs(vec3{-0.175f,-0.1f,-0.5f}, quat_from_angles(0,200,0), vec3{0.175f,0.175f,0.175f}));

	static pose_t window_pose =
		pose_t{ {0.175f,-0.05f,-0.4f}, quat_lookat({0.175f,-0.05f,-0.4f}, {0.175f,-0.05f,0}) };

	ui_window_begin("PostFX", &window_pose);

	if (ui_toggle("Vignette", vignette_on)) postfx_apply_chain();
	if (vignette_on && ui_hslider("strength", vignette_strength, 0, 1))
		material_set_float(postfx_vignette, "strength", vignette_strength);

	if (ui_toggle("Invert", invert_on)) postfx_apply_chain();

	if (ui_toggle("Depth fog", fog_on)) postfx_apply_chain();
	if (fog_on && ui_hslider("density", fog_density, 0, 2))
		material_set_float(postfx_fog, "fog_density", fog_density);

	ui_window_end();
}

///////////////////////////////////////////

void demo_postfx_shutdown() {
	render_set_post_process(nullptr, 0);
	vignette_on = false;
	invert_on   = false;
	fog_on      = false;

	material_release(postfx_vignette);
	material_release(postfx_invert);
	material_release(postfx_fog);
	shader_release  (postfx_vignette_shader);
	shader_release  (postfx_invert_shader);
	shader_release  (postfx_fog_shader);
	model_release   (postfx_model);
}

File: Examples/StereoKitTest/Demos/DemoPostFX.cs
// SPDX-License-Identifier: MIT
// The authors below grant copyright rights under the MIT license:
// Copyright (c) 2026 Nick Klingensmith
// Copyright (c) 2026 Qualcomm Technologies, Inc.

using StereoKit;
using System.Collections.Generic;

class DemoPostFX : ITest
{
	string title       = "Post Processing";
	string description = "On mobile hardware, post processing can be VERY expensive! StereoKit supports 'subpass' based post processing, accepting some limitations to enable a subset of post processing effects that do run fast (in-tile) on mobile hardware!";

	// Helmet and settings sit side by side, centered together in front.
	Pose windowPose = Matrix.T(0.175f, 0.1f, 0) * Demo.contentPose.Pose;

	Material vignetteMat;
	Material invertMat;
	Material fogMat;
	bool     vignetteOn;
	bool     invertOn;
	bool     fogOn;
	float    vignetteStrength = 0.4f;
	float    fogDensity       = 0.5f;

	Model model;

	public void Initialize()
	{
		/// :CodeSample: Renderer.SetPostProcess
		/// ### Setting a post-process chain
		/// A post-process effect is just a Material! Its shader reads the
		/// scene through an input attachment named 'color', and its
		/// parameters can be changed live, like any other Material. The
		/// chain renders in argument order.
		Material vignette = new Material("postfx_vignette.hlsl");
		vignette["strength"] = 0.4f;
		Renderer.SetPostProcess(vignette);
		///
		/// And when you're done with it:
		Renderer.SetPostProcess();
		/// :End:
		vignetteMat = vignette;

		invertMat = new Material("postfx_math.hlsl");
		invertMat["mul_color"] = new Vec3(-1, -1, -1);
		invertMat["add_color"] = new Vec3( 1,  1,  1);

		fogMat = new Material("postfx_depth_fog.hlsl");
		fogMat["fog_density"] = fogDensity;

		/// :CodeSample: RenderSettings Renderer.RenderTo
		/// ### Rendering a viewpoint with post-processing
		/// RenderSettings works with Renderer.RenderTo and RenderList.DrawNow,
		/// and can carry a post-process chain that applies to just that pass!
		Tex target = Tex.RenderTarget(512, 512);
		Renderer.RenderTo(target, Matrix.T(0, 0, 1), Matrix.Perspective(90, 1, 0.1f, 50),
			new RenderSettings { clearColor  = Color.Black,
			                     postProcess = new Material[] { vignette } });
		/// :End:

		model = Model.FromFile("DamagedHelmet.gltf");

		// Testing takes a screenshot of this demo, and screenshots follow the
		// display's post-process chain - turn some effects on for the shot!
		if (Tests.IsTesting)
		{
			vignetteOn = true;
			fogOn      = true;
			ApplyChain();
		}
	}

	// Chain order is the array order - fog reads scene depth first, then the
	// invert flips colors, and the vignette darkens the corners last.
	void ApplyChain()
	{
		List<Material> chain = new List<Material>();
		if (fogOn     ) chain.Add(fogMat);
		if (invertOn  ) chain.Add(invertMat);
		if (vignetteOn) chain.Add(vignetteMat);
		Renderer.SetPostProcess(chain.ToArray());
	}

	public void Shutdown()
	{
		Renderer.SetPostProcess();
	}

	public void Step()
	{
		UI.WindowBegin("Post Processing", ref windowPose, new Vec2(0.28f, 0));

		if (UI.Toggle("Vignette", ref vignetteOn)) ApplyChain();
		if (vignetteOn && UI.HSlider("strength", ref vignetteStrength, 0, 1))
			vignetteMat["strength"] = vignetteStrength;

		if (UI.Toggle("Invert", ref invertOn)) ApplyChain();

		if (UI.Toggle("Depth fog", ref fogOn)) ApplyChain();
		if (fogOn && UI.HSlider("density", ref fogDensity, 0, 2))
			fogMat["fog_density"] = fogDensity;

		UI.WindowEnd();

		// Left of the settings window, turned to a 3/4 view (contentPose already
		// yaws 180, so 20 here lands the helmet at a 200 world yaw).
		model.Draw(Matrix.TRS(V.XYZ(0.175f, 0, -0.1f), Quat.FromAngles(0, 20, 0), 0.175f) * Demo.contentPose);

		Tests.Screenshot("Demos/PostFX.jpg", 600, 600, V.XYZ(-0.175f, -0.1f, -0.08f), V.XYZ(-0.175f, -0.1f, -0.5f));
		Demo.ShowSummary(title, description, new Bounds(V.XY0(0, -0.02f), V.XYZ(.65f, .45f, 0.6f)));
	}
}
